### ISSUE TITLE
Sequence Generator 2.0

### ISSUE THREAD

**Issue by rizar** [mila-udem/blocks#988]
I am working on the second version of the sequence generation functionality for Blocks. I would appreciate feedback and suggestions, especially from @janchorowski , @dmitriy-serdyuk @sotelo @orhanf , cause I know you guys have had experience with this code.

My main goals for the new version are the following:
- Hide the attention mechanism into the recurrent transition. I think this can be achieved if at every step we first generate the next output token, then compute new states of the RNN, then do the attention. 
- Reduce conceptual overload by having fewer and less abstract components. For example, I want to get rid of this crazy two level nested structure, when `SequenceGenerator` consists of a few components, which in turn, e.g. `Readout`, are composite. I am hesitating now between two options:
    - a sequence generator can be composed of a recurrent brick and a readout
    - a sequence generator can be composed of a recurrent brick, readout, and a feedback brick
So far it seems like the second option is the way to go, because the details of how the generated symbols are fed into the recurrent brick are rather orthogonal two the process of generation
- Move the computation of the cost into `Readout` and make it a function of both `prediction` and `groundtruth` sequences. This is required to implement advanced training methods such as REINFORCE.

In a nutshell, the idea is to make it more compact and more powerful at the same time. There must be a cost for this, and so far I remember at least one feature that does not fit this picture: native support of direct dependency of the generated symbol on the previous one. This hack costs too much code for the moderate benefit it brings. 

Let me know your thoughts! As soon as I have anything working, I will open a PR to Blocks-extras.

**Comment by rizar** [comment]
You can track  the progress [here](https://github.com/mila-udem/blocks-extras/pull/43).
